engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeAnyNode.java

```diff
@@ -0,0 +1,349 @@
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.callable.atom.AtomConstructor;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.error.WarningsLibrary;
+import org.enso.interpreter.runtime.number.EnsoBigInteger;
+
+/**
+ * Implements {@code hash_code} functionality.
+ *
+ * <p>Make sure that the hashing contract is retained after any modification.
+ *
+ * <h3>Hashing contract:</h3>
+ *
+ * <ul>
+ *   <li>Whenever two objects are equal ({@code EqualsAnyNode} returns {@code true}), their hashcode
+ *       should equal. More formally: {@code For all objects o1, o2: if o1 == o2 then hash(o1) ==
+ *       hash(o2)}
+ *   <li>Whenever two hash codes are different, their associated objects are different: {@code For all objects
+ *       o1, o2: if hash(o1) != hash(o2) then o1 != o2.
+ * </ul>
+ */
+@BuiltinMethod(type = "Any", name = "hash_code", description = "Returns hash code of the object.")
+@GenerateUncached
+public abstract class HashCodeAnyNode extends Node {
+
+  public static HashCodeAnyNode build() {
+    return HashCodeAnyNodeGen.create();
+  }
+
+  public abstract long execute(@AcceptsError Object self);
+
+  /** Specializations for primitive values * */
+  @Specialization
+  long hashCodeForShort(short s) {
+    return s;
+  }
+
+  @Specialization
+  long hashCodeForByte(byte b) {
+    return b;
+  }
+
+  @Specialization
+  long hashCodeForLong(long l) {
+    return l;
+  }
+
+  @Specialization
+  long hashCodeForInt(int i) {
+    return i;
+  }
+
+  @Specialization
+  long hashCodeForFloat(float f) {
+    return Float.hashCode(f);
+  }
+
+  @Specialization
+  long hashCodeForDouble(double d) {
+    return d % 1.0 == 0.0 ? (int) d : Double.hashCode(d);
+  }
+
+  @Specialization
+  @TruffleBoundary
+  long hashCodeForBigInteger(EnsoBigInteger bigInteger) {
+    return bigInteger.getValue().hashCode();
+  }
+
+  @Specialization
+  long hashCodeForAtomConstructor(AtomConstructor atomConstructor) {
+    return System.identityHashCode(atomConstructor);
+  }
+
+  /** How many {@link HashCodeAnyNode} nodes should be created for fields in atoms. */
+  static final int hashCodeNodeCountForFields = 10;
+
+  static HashCodeAnyNode[] createHashCodeNodes(int size) {
+    HashCodeAnyNode[] nodes = new HashCodeAnyNode[size];
+    Arrays.fill(nodes, HashCodeAnyNode.build());
+    return nodes;
+  }
+
+  @Specialization
+  long hashCodeForAtom(
+      Atom atom,
+      @Cached(value = "createHashCodeNodes(hashCodeNodeCountForFields)", allowUncached = true)
+          HashCodeAnyNode[] fieldHashCodeNodes,
+      @Cached ConditionProfile isHashCodeCached,
+      @Cached ConditionProfile enoughHashCodeNodesForFields,
+      @Cached LoopConditionProfile loopProfile) {
+    if (isHashCodeCached.profile(atom.getHashCode() != null)) {
+      return atom.getHashCode();
+    }
+    // TODO[PM]: If atom overrides hash_code, call that method
```

Comment:
?

Reply:
This is just a side node for myself, once I will be working on https://www.pivotaltracker.com/story/show/183945328, which will be immediate follow-up of this PR. It means that if the `atom` provides an overriden `hash_code` (probably in form of overriding `hash_code` method on its `Comparator`), we should call that, instead of computing it. In a similar approach to what is done in https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAnyNode.java#L402 . But I understand that this note is rather unreadable. I promise that I will get rid of it once I work on the follow-up PR.

Reply:
Ah, so that will be implemented in follow up work. Sounds good. Just reference the ticket in that comment and it's all good.